File: frontend/src/api/aiSuggestion.ts
import { get, post, put } from '@/utils/request'

const PREFIX = '/api/v1/ai-suggestion'

export type AISuggestionMode = 'manual' | 'suggestion' | 'auto'

export interface AISuggestionAccountSetting {
  account_id: string
  mode: AISuggestionMode
  profile_id: number | null
  inherited_profile: boolean
  review_delay_ms: number
  inherit_review_delay: boolean
  reply_style: {
    tone: 'professional' | 'friendly' | 'concise' | 'warm'
    form_of_address: string
    length: 'short' | 'medium' | 'detailed'
    use_emoji: boolean
  }
  inherit_reply_style: boolean
  custom_prompt: string
}

export interface AIGroupMessage {
  role: 'buyer' | 'seller'
  content: string
  source_message_id?: string
}

export interface AIBusinessContext {
  item_title?: string
  item_description?: string
  item_price?: string
}

export interface AISuggestionResult {
  record_id: number
  suggestion: string
  provider_name: string
  model_name: string
  latency_ms: number
}

export interface AIConnectionProfile {
  id: number
  name: string
  provider_type: 'deepseek' | 'openai_compatible'
  base_url: string
  model_name: string
  enabled: boolean
  is_global_default: boolean
  has_api_key: boolean
  allowed_user_ids: number[] | null
  fallback_profile_ids: number[] | null
  input_price_per_million: string | null
  output_price_per_million: string | null
  created_at: string
  updated_at: string
}

export interface AIProfileInput {
  name: string
  provider_type: 'deepseek' | 'openai_compatible'
  base_url: string
  model_name: string
  api_key?: string
  enabled: boolean
  is_global_default: boolean
  fallback_profile_ids: number[]
  input_price_per_million?: number
  output_price_per_million?: number
}

export interface AISuggestionGlobalSetting {
  review_delay_ms: number
  reply_style: AISuggestionAccountSetting['reply_style']
  custom_prompt: string
}

export interface AISuggestionRecordRow {
  id: number
  account_id: string
  conversation_id: string
  status: 'generated' | 'sent' | 'edited_sent' | 'ignored' | 'failed'
  provider_type: string | null
  model_name: string | null
  prompt_tokens: number | null
  completion_tokens: number | null
  total_tokens: number | null
  latency_ms: number | null
  estimated_cost: string | null
  created_at: string
}

export interface AISuggestionSummaryRow {
  day: string
  account_id: string
  profile_id: number | null
  provider_type: string | null
  model_name: string | null
  request_count: number
  total_tokens: number
  estimated_cost: string
}

interface ApiResult<T = unknown> {
  success: boolean
  message?: string
  data?: T
}

export const getAISuggestionAccountSetting = async (accountId: string) => {
  const res = await get<ApiResult<AISuggestionAccountSetting>>(`${PREFIX}/accounts/${encodeURIComponent(accountId)}/settings`)
  if (!res.success || !res.data) throw new Error(res.message || '获取 AI 建议设置失败')
  return res.data
}

export const generateAISuggestion = (
  accountId: string,
  conversationId: string,
  groupId: string,
  messages: AIGroupMessage[],
  businessContext?: AIBusinessContext,
  regenerateInstruction?: string,
) => post<ApiResult<AISuggestionResult & { blocked?: boolean; risks?: Array<{ message_index: number; types: string[] }> }>>(
  `${PREFIX}/generate`,
  {
    account_id: accountId,
    conversation_id: conversationId,
    group_id: groupId,
    messages,
    business_context: businessContext && Object.values(businessContext).some(Boolean)
      ? businessContext
      : undefined,
    regenerate_instruction: regenerateInstruction || undefined,
  },
)

export const rejectAIMessageGroup = (accountId: string, conversationId: string, groupId: string) =>
  post<ApiResult>(`${PREFIX}/groups/reject`, {
    account_id: accountId,
    conversation_id: conversationId,
    group_id: groupId,
  })

export const updateAISuggestionAction = (
  recordId: number,
  action: 'sent' | 'edited_sent' | 'ignored',
  finalContent?: string,
) => put<ApiResult<{ context_saved: boolean }>>(`${PREFIX}/records/${recordId}/action`, {
  action,
  final_content: finalContent,
})

export const listAIConnectionProfiles = async () => {
  const res = await get<ApiResult<AIConnectionProfile[]>>(`${PREFIX}/profiles`)
  if (!res.success) throw new Error(res.message || '获取 AI 连接配置失败')
  return res.data || []
}

export const createAIConnectionProfile = (payload: AIProfileInput & { api_key: string }) =>
  post<ApiResult<AIConnectionProfile>>(`${PREFIX}/profiles`, payload)

export const updateAIConnectionProfile = (profileId: number, payload: AIProfileInput) =>
  put<ApiResult<AIConnectionProfile>>(`${PREFIX}/profiles/${profileId}`, payload)

export const testAIConnectionProfile = (profileId: number) =>
  post<ApiResult<{ model_name: string; latency_ms: number }>>(`${PREFIX}/profiles/${profileId}/test`, {})

export const getAISuggestionGlobalSetting = async () => {
  const res = await get<ApiResult<AISuggestionGlobalSetting>>(`${PREFIX}/global-settings`)
  if (!res.success || !res.data) throw new Error(res.message || '获取全局设置失败')
  return res.data
}

export const updateAISuggestionGlobalSetting = (payload: AISuggestionGlobalSetting) =>
  put<ApiResult<AISuggestionGlobalSetting>>(`${PREFIX}/global-settings`, payload)

export const saveAISuggestionAccountSetting = (
  accountId: string,
  payload: Omit<AISuggestionAccountSetting, 'account_id' | 'inherited_profile'>,
) => put<ApiResult<AISuggestionAccountSetting>>(`${PREFIX}/accounts/${encodeURIComponent(accountId)}/settings`, payload)

export const listAISuggestionRecords = async (accountId?: string) => {
  const query = accountId ? `?account_id=${encodeURIComponent(accountId)}` : ''
  const res = await get<ApiResult<AISuggestionRecordRow[]>>(`${PREFIX}/records${query}`)
  if (!res.success) throw new Error(res.message || '获取建议记录失败')
  return res.data || []
}

export const summarizeAISuggestionRecords = async () => {
  const res = await get<ApiResult<AISuggestionSummaryRow[]>>(`${PREFIX}/records/summary`)
  if (!res.success) throw new Error(res.message || '获取建议统计失败')
  return res.data || []
}

File: common/schemas/ai_suggestion.py
"""AI 建议模式 API 数据结构。"""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field, HttpUrl, field_validator


ProviderType = Literal["deepseek", "openai_compatible"]
SuggestionMode = Literal["manual", "suggestion", "auto"]


class AIConnectionProfileCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    provider_type: ProviderType
    base_url: str = Field(default="", max_length=500)
    model_name: str = Field(min_length=1, max_length=160)
    api_key: str = Field(min_length=1, max_length=4000, repr=False)
    enabled: bool = True
    is_global_default: bool = False
    allowed_user_ids: list[int] | None = None
    fallback_profile_ids: list[int] | None = None
    input_price_per_million: Decimal | None = Field(default=None, ge=0)
    output_price_per_million: Decimal | None = Field(default=None, ge=0)

    @field_validator("base_url")
    @classmethod
    def validate_base_url(cls, value: str) -> str:
        value = value.strip().rstrip("/")
        if value and not value.startswith(("https://", "http://")):
            raise ValueError("Base URL 必须以 http:// 或 https:// 开头")
        return value


class AIConnectionProfileUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    provider_type: ProviderType | None = None
    base_url: str | None = Field(default=None, max_length=500)
    model_name: str | None = Field(default=None, min_length=1, max_length=160)
    api_key: str | None = Field(default=None, min_length=1, max_length=4000, repr=False)
    enabled: bool | None = None
    is_global_default: bool | None = None
    allowed_user_ids: list[int] | None = None
    fallback_profile_ids: list[int] | None = None
    input_price_per_million: Decimal | None = Field(default=None, ge=0)
    output_price_per_million: Decimal | None = Field(default=None, ge=0)


class AIConnectionProfileView(BaseModel):
    id: int
    name: str
    provider_type: ProviderType
    base_url: str
    model_name: str
    enabled: bool
    is_global_default: bool
    has_api_key: bool
    allowed_user_ids: list[int] | None = None
    fallback_profile_ids: list[int] | None = None
    input_price_per_million: Decimal | None = None
    output_price_per_million: Decimal | None = None
    created_at: datetime
    updated_at: datetime


class ReplyStyle(BaseModel):
    tone: Literal["professional", "friendly", "concise", "warm"] = "friendly"
    form_of_address: str = Field(default="亲", max_length=40)
    length: Literal["short", "medium", "detailed"] = "short"
    use_emoji: bool = False


class AISuggestionAccountSettingUpdate(BaseModel):
    mode: SuggestionMode = "manual"
    profile_id: int | None = None
    review_delay_ms: int = Field(default=4000, ge=1000, le=30000)
    inherit_review_delay: bool = True
    reply_style: ReplyStyle = Field(default_factory=ReplyStyle)
    inherit_reply_style: bool = True
    custom_prompt: str = Field(default="", max_length=4000)


class AISuggestionAccountSettingView(AISuggestionAccountSettingUpdate):
    account_id: str
    inherited_profile: bool = True


class AISuggestionGlobalSetting(BaseModel):
    review_delay_ms: int = Field(default=4000, ge=1000, le=30000)
    reply_style: ReplyStyle = Field(default_factory=ReplyStyle)
    custom_prompt: str = Field(default="", max_length=4000)


class AIGroupMessage(BaseModel):
    role: Literal["buyer", "seller"]
    content: str = Field(min_length=1, max_length=12000)
    source_message_id: str | None = Field(default=None, max_length=160)


class AIBusinessContext(BaseModel):
    item_title: str | None = Field(default=None, max_length=500)
    item_description: str | None = Field(default=None, max_length=4000)
    item_price: str | None = Field(default=None, max_length=80)
    order_status: str | None = Field(default=None, max_length=120)


class AISuggestionGenerateRequest(BaseModel):
    account_id: str = Field(min_length=1, max_length=80)
    conversation_id: str = Field(min_length=1, max_length=128)
    group_id: str = Field(min_length=1, max_length=64)
    messages: list[AIGroupMessage] = Field(min_length=1, max_length=50)
    business_context: AIBusinessContext | None = None
    regenerate_instruction: str | None = Field(default=None, max_length=1000)


class AIGroupRejectRequest(BaseModel):
    account_id: str = Field(min_length=1, max_length=80)
    conversation_id: str = Field(min_length=1, max_length=128)
    group_id: str = Field(min_length=1, max_length=64)


class AISuggestionActionRequest(BaseModel):
    action: Literal["sent", "edited_sent", "ignored"]
    final_content: str | None = Field(default=None, max_length=12000)
